airflow/contrib/hooks/sagemaker_hook.py

```diff
@@ -16,299 +16,793 @@
     """
-    non_terminal_states = {'InProgress', 'Stopping', 'Stopped'}
+    non_terminal_states = {'InProgress', 'Stopping'}
+    endpoint_non_terminal_states = {'Creating', 'Updating', 'SystemUpdating',
+                                    'RollingBack', 'Deleting'}
     failed_states = {'Failed'}
 
     def __init__(self,
-                 sagemaker_conn_id=None,
-                 use_db_config=False,
-                 region_name=None,
-                 check_interval=5,
-                 max_ingestion_time=None,
                  *args, **kwargs):
         super(SageMakerHook, self).__init__(*args, **kwargs)
-        self.sagemaker_conn_id = sagemaker_conn_id
-        self.use_db_config = use_db_config
-        self.region_name = region_name
-        self.check_interval = check_interval
-        self.max_ingestion_time = max_ingestion_time
-        self.conn = self.get_conn()
+        self.s3_hook = S3Hook(aws_conn_id=self.aws_conn_id)
+
+    def expand_role(self, role):
+        """
+        Expand an IAM role name to an IAM role ARN. If role is already an IAM ARN,
+        no change is made.
+
+        :param role: IAM role name or ARN
+        :return: IAM role ARN
+        """
+        if '/' in role:
+            return role
+        else:
+            return self.get_iam_conn().get_role(RoleName=role)['Role']['Arn']
+
+    def tar_and_s3_upload(self, path, key, bucket):
+        """
+        Tar the local file or directory and upload to s3
 
-    def check_for_url(self, s3url):
+        :param path: local file or directory
+        :type path: str
+        :param key: s3 key
+        :type key: str
+        :param bucket: s3 bucket
+        :type bucket: str
+        :return: None
+        """
+        with tempfile.TemporaryFile() as temp_file:
+            if os.path.isdir(path):
+                files = [os.path.join(path, name) for name in os.listdir(path)]
+            else:
+                files = [path]
+            with tarfile.open(mode='w:gz', fileobj=temp_file) as tar_file:
+                for f in files:
+                    tar_file.add(f, arcname=os.path.basename(f))
+            temp_file.seek(0)
+            self.s3_hook.load_file_obj(temp_file, key, bucket, True)
+
+    def configure_s3_resources(self, config):
+        """
+        Extract the S3 operations from the configuration and execute them.
+
+        :param config: config of SageMaker operation
+        :type config: dict
+        :return: dict
         """
-        check if the s3url exists
+        s3_operations = config.pop('S3Operations', None)
+
+        if s3_operations is not None:
+            create_bucket_ops = s3_operations.get('S3CreateBucket')
+            upload_ops = s3_operations.get('S3Upload')
+            if create_bucket_ops:
+                for op in create_bucket_ops:
+                    self.s3_hook.create_bucket(bucket_name=op['Bucket'])
+            if upload_ops:
+                for op in upload_ops:
```

Comment:
```suggestion
            for op in s3_operations.get('S3Upload', []):
```

and we can get rid of the `if` too. (I generally prefer to nest things as little as possible as a general rule, as it I find it makes it easier to understand.

Reply:
Updated.